### PR TITLE
Implement dumps

### LINKED ISSUE #35: Implement dumps (add tests and code samples)
Since [v0.15.0](https://github.com/meilisearch/MeiliSearch/releases/tag/v0.15.0), MeiliSearch exposes a new functionality: [dumps](https://docs.meilisearch.com/references/dump.html#dumps)

This feature should be implemented in this SDK following [this specificacions](https://github.com/meilisearch/integration-guides/issues/44)

- [ ] Implement dump creation route
- [ ] Implement get dump status route
- [ ] Add tests
- [ ] Add code samples

### PR DESCRIPTION
**Pull request by luis-valdez**
## General info
Closes #35.
This PR implements the dumps route of the MeiliSearch API. 
For more details about dumps: https://docs.meilisearch.com/references/dump.html

### Implement dump creation route

 The new class `Dump` works as a wrapper for `MeiliSearchRequestHttp` and handles the creation of the dump with the method `createDump`. This method returns a Dump object which contains:
- `uid` that is the unique identifier of the dump
- `status` which indicates the status of the dump at the moment of the call

The method `createDump` is finally called by the `Client` class in a method of its own, that also returns an instance of Dump.

###  Implement get dump status route
Very similar to the previous point, it uses the new `Dump` class in coordination with the `Client` class to make an HTTP GET request and retrieve the current information of the dump. It uses the method `getStatusByUid(String uid)` in the Dump class and the `getDumpStatus(String uid)` in the Client class.
An example return value of `getDumpStatus`
```
{"uid":"20201218-023314939","status":"done"}
```
 
 ### Add tests
To validate the previous 2 points, two new tests were implemented in  the `ClientTest` class. This tests are `testGetDumpStatus` and `testCreateDump`.
1. `testCreateDump` creates a dump and validates if it has a status of `in_progress`
2. `testGetDumpStatus` creates a dump and gets its uid. Then it validates if it can retrieve the status using `getStatusByUid`

 ### Add code samples
In this part I wasn't sure where should I include the code samples. But here they are:
```
import com.meilisearch.sdk.Client;
import com.meilisearch.sdk.Config;
import com.meilisearch.sdk.Index;
import com.meilisearch.sdk.Dump;

class TestMeiliSearch {
  public static void main(String[] args) throws Exception {

    final String documents = "["
      + "{\"book_id\": 123, \"title\": \"Pride and Prejudice\"},"
      + "{\"book_id\": 456, \"title\": \"Le Petit Prince\"},"
      + "{\"book_id\": 1, \"title\": \"Alice In Wonderland\"},"
      + "{\"book_id\": 1344, \"title\": \"The Hobbit\"},"
      + "{\"book_id\": 4, \"title\": \"Harry Potter and the Half-Blood Prince\"},"
      + "{\"book_id\": 2, \"title\": \"The Hitchhiker\'s Guide to the Galaxy\"}"
      + "]";

    Client client = new Client(new Config("http://localhost:7700", "masterKey"));
    
    // An index is where the documents are stored.
    Index index = client.index("books");

    // If the index 'books' does not exist, MeiliSearch creates it when you first add the documents.
    index.addDocuments(documents); // => { "updateId": 0 }
    
    Dump dump = client.createDump();
    String status = dump.getStatus(); // in_progress
    String dumpUid = dump.getUid()); // 20201218-023314939
    String statusAfterCreation = client.getDumpStatus(dumpUid); // {"uid":"20201218-023314939","status":"done"}
  }
}
```

If you guys have any questions regarding this PR, please let me know. And if you have any feedback is very appreciated 😄!

### CODE REVIEW

**Comment by curquiza** [comment]
Hello @luis-valdez!
Thanks again for your involvement! As @eskombro said, the holiday season will start for us soon, so we won't be available for a moment.
Btw, I've just added you to the repository with write access since you are really involved! 😄 Now you don't have to work on your forked repository anymore. More convenient! 😉

**Comment by curquiza** [comment]
We also noticed you don't appear in the contributors of this repository despite your several PRs
Capture d’écran 2020-12-18 à 18 32 13

I've investigated, and it might be your commits that are not linked to the address email of your account. It means GitHub cannot know your GitHub account [luis-valdez](https://github.com/luis-valdez) is linked to your commits. We think you highly deserve appearing in our contributor list, so if you are ok with it, you might want to set your email to your git config:
```bash
$ git config --global user.email "MY_NAME@example.com"
```
😁

**Comment by luis-valdez** [comment]
> We also noticed you don't appear in the contributors of this repository despite your several PRs
> I've investigated, and it might be your commits that are not linked to the address email of your account. It means GitHub cannot know your GitHub account [luis-valdez](https://github.com/luis-valdez) is linked to your commits. We think you highly deserve appearing in our contributor list, so if you are ok with it, you might want to set your email to your git config:
> 
> ```shell
> $ git config --global user.email "MY_NAME@example.com"
> ```
> 
> 😁

Hello @curquiza, first of all, thank you a lot for adding me as a collaborator 😄!
I noticed that same issue too! I believe I already fixed it and when my next PR is accepted I'll appear as one of the collaborators. 

Thank you a lot for letting me contribute, and happy holidays 🎉 !

File: src/test/java/com/meilisearch/integration/ClientTest.java [hunk 1]
```diff
@@ -209,4 +210,25 @@ public void testIndexMethodCallExistingIndexWithPrimaryKey() throws Exception {
...
 	}
 
+	/**
+	 * Test call to create dump
+	 */
+	@Test
+	public void testCreateDump() throws Exception {
+		Dump dump = client.createDump();
+		String status = dump.getStatus();
+		System.out.println(dump.getUid());
```
Comment (eskombro) [inline src/test/java/com/meilisearch/integration/ClientTest.java:220]:
```suggestion
```

File: src/main/java/com/meilisearch/sdk/Dump.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
+package com.meilisearch.sdk;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
```
Comment (eskombro) [inline src/main/java/com/meilisearch/sdk/Dump.java:4]:
```suggestion
```

File: src/main/java/com/meilisearch/sdk/Dump.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	 * @return Dump object with Meilisearch API response
+	 * @throws Exception if something goes wrong
+	 */
+	public Dump createDump() throws Exception, MeiliSearchApiException {
+		Dump dump = new Gson().fromJson(
+			this.meiliSearchHttpRequest.post("/dumps", ""),
+			Dump.class
+		);
+
+		return dump;
```
Comment (eskombro) [inline src/main/java/com/meilisearch/sdk/Dump.java:35]:
```suggestion
		return new Gson().fromJson(
			this.meiliSearchHttpRequest.post("/dumps", ""),
			Dump.class
		);
```

Comment (eskombro) [reply src/main/java/com/meilisearch/sdk/Dump.java:35]: As a suggestion: you can directly return the value. If you are not manipulating it in any way, there is no need to store it in a variable :)

**Review by eskombro — CHANGES_REQUESTED** [review]
Hey @luis-valdez !

I'm back and checking your work, thanks! I see that this is working as expected, and that is a nice starting point! Do not worry about code samples, we can omit this for this PRas no code examples have been done at all for the Java SDK and this should be fixed by https://github.com/meilisearch/meilisearch-java/issues/20 at some point.

Basically, the way this works seems to fulfill the goal and tests are good. But I get a feeling that there is something a little strange in the structure of what a Dump is (in terms of the codebase). I feel that the Dump class is used to englobe two different things: on one side, a Dump status (which is why it has properties as `status` and `uid`), but also, as some kind of "dumps handler".

This confusion seems more evident to me when you write [this](https://github.com/meilisearch/meilisearch-java/pull/89/files#diff-a0c212354bad44687d6bdf50427d50cab5a7d6e9252d5851396fc79ca2eb2ec7R166):

```java
this.dump.createDump();
```

When I read it, I see the confusion. Is the dump attribute a dump in itself? Is it a handler that creates a dump object? That line, idiomatically, is asking a dump to create a dump :)

So my `client.dump` property doesn't represent a MeiliSearch `Dump` itself. But the result of `client.createDump()` does. Yet they are both of the same type (class) and expose the same methods.

What do you think if we do two separate classes for these two different functionalities? For example, a `Dump` class which contains the information of a single Dump, and a `DumpHandler` which belongs to the client and exposes the methods `createDump()` or `getDumpStatus()`?

Let me know what you think!

I also suggested removing two lines that don't seem to be used anymore :)

**Comment by luis-valdez** [comment]
Hey @eskombro! Glad to see you back.
You're right, I was violating the Single Responsibility Principle, and I wasn't noticing.
I implemented the `DumpHandler` class. And refactor the code to meet your requests.

Thank you a lot for your feedback, and let me know if there's anything else you think can be improved 😄

File: src/test/java/com/meilisearch/integration/ClientTest.java [hunk 4]
```diff
@@ -209,4 +210,24 @@ public void testIndexMethodCallExistingIndexWithPrimaryKey() throws Exception {
...
+
+	/**
+	 * Test call to get dump status by uid
+	 */
+	@Test
+	public void testGetDumpStatus() throws Exception {
+		Dump dump = client.createDump();
+		String uid = dump.getUid();
+		String status = client.getDumpStatus(uid);
+		assertNotNull(status);
```
Comment (eskombro) [inline src/test/java/com/meilisearch/integration/ClientTest.java:231]:
```suggestion
		assertNotNull(status);
		assertNotNull(uid);
```

**Review by eskombro — APPROVED** [review]
Thanks a lot for the changes @luis-valdez !

I am really happy with the result, and when I read the code it seems more clear!

I also realized that your branch was based on the previous one, but it was too late! I don't think we will have any conflicts though, but if you prefer you can rebase your branch on master and push-force it, then we review and merge :)

But it looks good to me as it is!

**Comment by eskombro** [comment]
bors try

**Comment by luis-valdez** [comment]
Yes I agree, the code is way more understandable now. Great code review!

If you think it'll be okay then I'm good with merging it as it is.

**Comment by eskombro** [comment]
It's safe, this is what will be committed:

https://github.com/meilisearch/meilisearch-java/commit/2ef9d2ad8444c7ac47c71edca6a7a4237251edb3

So no worries @luis-valdez, thanks for this PR!!! 🎉 

bors merge

**Review by eskombro — APPROVED** [review]
🚀

**Comment by eskombro** [comment]
bors merge

**Pull request merged**